analyzers/src/SonarAnalyzer.CFG/ShimLayer/AnalysisContext/SymbolStartAnalysisContext.cs

```diff
@@ -0,0 +1,169 @@
+        }
+        else
+        {
+            throw new ArgumentException("Invalid type parameter.", nameof(TLanguageKindEnum));
+        }
+    }
+
+    public void RegisterOperationAction(Action<OperationAnalysisContext> action, ImmutableArray<OperationKind> operationKinds) =>
+        registerOperationAction(RoslynSymbolStartAnalysisContext, action, operationKinds);
+
+    public void RegisterOperationBlockAction(Action<OperationBlockAnalysisContext> action) =>
+        registerOperationBlockAction(RoslynSymbolStartAnalysisContext, action);
+
+    public void RegisterOperationBlockStartAction(Action<OperationBlockStartAnalysisContext> action) =>
+        registerOperationBlockStartAction(RoslynSymbolStartAnalysisContext, action);
+
+    public void RegisterSymbolEndAction(Action<SymbolAnalysisContext> action) =>
+        registerSymbolEndAction(RoslynSymbolStartAnalysisContext, action);
+
+    public void RegisterSyntaxNodeAction<TLanguageKindEnum>(Action<SyntaxNodeAnalysisContext> action, params TLanguageKindEnum[] syntaxKinds) where TLanguageKindEnum : struct
+    {
+        var languageKindType = typeof(TLanguageKindEnum);
+        if (languageKindType == typeof(CS.SyntaxKind))
+        {
+            registerSyntaxNodeActionCS(RoslynSymbolStartAnalysisContext, action, syntaxKinds.Cast<CS.SyntaxKind>().ToImmutableArray());
+        }
+        else if (languageKindType.FullName == "Microsoft.CodeAnalysis.VisualBasic.SyntaxKind")
+        {
+            throw new NotImplementedException("Add a reference to the Microsoft.CodeAnalysis.VisualBasic.Workspaces package.");
+        }
+        else
+        {
+            throw new ArgumentException("Invalid type parameter.", nameof(TLanguageKindEnum));
+        }
+    }
+
+    // symbolStartAnalysisContextParameter => ((symbolStartAnalysisContextType)symbolStartAnalysisContextParameter)."propertyName"
+    private static Func<object, TProperty> CreatePropertyAccessor<TProperty>(Type symbolStartAnalysisContextType, string propertyName)
+    {
+        var symbolStartAnalysisContextParameter = Parameter(typeof(object));
+        return Lambda<Func<object, TProperty>>(
+            Property(
+                Convert(symbolStartAnalysisContextParameter, symbolStartAnalysisContextType), propertyName),
+            symbolStartAnalysisContextParameter).Compile();
+    }
+
+    // (object symbolStartAnalysisContextParameter, Action<TContext> registerActionParameter)
+    //     => ((symbolStartAnalysisContextType)symbolStartAnalysisContextParameter)."registrationMethodName"<typeArguments>(registerActionParameter)
+    private static Action<object, Action<TContext>> CreateRegistrationMethod<TContext>(Type symbolStartAnalysisContextType, string registrationMethodName, params Type[] typeArguments)
+    {
+        var receiver = Parameter(typeof(object));
+        var registerActionParameter = Parameter(typeof(Action<TContext>));
+        return Lambda<Action<object, Action<TContext>>>(
```

Comment:
This is a case where I would put one parameter per line, on the outer function (i.e. `Lambda`):
```cs
return Lambda<Action<object, Action<TContext>>>(
    Call(Convert(receiver, symbolStartAnalysisContextType), registrationMethodName, typeArguments, registerActionParameter),
    receiver,
    registerActionParameter).Compile();
```

That makes it clearer what parameter belong to what, that is particularly important in the specific case, since the `ParameterExpression` comes last in `Lambda`, but should be before body if the syntactical order of a method is followed.